book/src/htlc.md

```diff
@@ -0,0 +1,74 @@
+# Hashed Time-Locked Contract Transactions
+
+A Hashed Time-Locked Contract is a transaction that allows one party of a
+transaction to spend tokens by forced-disclosure of the preimage of a message
+digest, and allows the counterparty to spend the tokens after a predefined
+lockout period.
+
+Another way to look at HTLC is that it is a protocol to eliminate [counterparty
+risk] where one side of a trade could default on their obligation.
+
+The key requirements for HTLC is that a ledger supports time-locked transactions,
+meaning a transaction that won't be executed until at a later point in time. And
+conditional transactions with message digest instructions.
+
+## Implementation
+
+There are a couple of ways to implement HTLC in Solana. We can implement HTLC
+as a transaction type or as a set of instructions akin to `OP_CHECKSEQUENCEVERIFY`
+and `OP_SHA256` in bitcoin [Script].
+
+### New Instructions
+
+In order to be compatible with [BIP-199] and [ERC-1630] we have to support 2 message
+digest algorithms, SHA-256 and Keccak-256 (SHA-3). I recommend that we also implement
+Blake2 and Shake message digest algorithms.
+
+We could implement these two instructions in Solana as [system instructions] as follows:
+
+```rust,ignore
+pub enum SystemInstruction {
+  /// Evaluates to true if period_in_seconds becomes zero.
+  /// Counter doesn't start until transaction is confirmed.
+  CheckElapsed { period_in_seconds: u64 },
+
+  /// SHA-256 hashing instruction.
+  /// Evaluates to the SHA-256 message digest.
+  SHA256 { data: &[u8] }
```

Comment:
We'd either use Budget for an instruction like this or create new instruction processor. System is the lowest-level instruction processor, is statically linked into the validator and only contains instructions necessary to create other instruction processors.  All others should ultimately be compiled to BPF and deployed to the chain via a shared object.  Those instruction processors need to "play by the rules" whereas System can make exceptions to bootstrap the system.

Reply:
I'm a little confused by the terminology and the fundamental difference between system instructions and just instructions. From the code I couldn't deduce what precisely is the difference or the rules that non-system instructions have to play by.

Also could you help point me to the right direction by pointing out the delineation point at which we implement a new instruction and have people write software that gets compiled/transpiled to BPF?

Reply:
SystemInstructions are special-cased in MessageProcessor. They, for example, are the only place where a "program_id" (instruction processor pubkey) can be modified.  Technically, `SystemInstruction::Transfer` could be moved out.  Also, currently, `SystemInstruction::CreateAccount` is the only instruction that's allowed to allocate memory, because the BPF programs currently don't include an allocator. That'll change in the future.

Reply:
Here's the code that implements those special cases: https://github.com/solana-labs/solana/blob/master/runtime/src/message_processor.rs#L48

Reply:
And described in the book here: https://solana-labs.github.io/book/runtime.html#execution